### PR TITLE
datadog: Add tracer option to check error eligibility in the span

### PR DESCRIPTION
**Pull request by vaihtovirta**
## What was changed

Added an option to datadog tracer that allows to determine whether an error is eligible to be added to the DD span.

Similar functionality in gorm dd instrumentation: [WithErrorCheck](https://github.com/DataDog/dd-trace-go/blob/50ffac63fe850477042ce4c4bad8de9efe15142d/contrib/gorm.io/gorm.v1/option.go#L70-L77)

## Why?

Not all errors need to be reported to Datadog (intermediate retry error, validation error, etc).
Users can filter out such errors via specifying a custom function via the option.

## Checklist

1. How was this tested:
I've added simple unit tests.

2. Any docs updates needed?
I didn't find docs related to temporal datadog tracer in the repository, but potentially this page can mention the new option: https://docs.temporal.io/dev-guide/go/observability#useful-resources

### CODE REVIEW

File: contrib/datadog/tracing/interceptor.go
Comment (cretz) [inline contrib/datadog/tracing/interceptor.go:51]: `Fn` not an abbreviation we usually use. How about:

```suggestion
	// CheckError can be set to a custom function to determine if an error should be traced.
	CheckError func(error) bool
```

Or maybe even better and more flexible:

```suggestion
	// OnFinish sets finish options. If unset, this will use [tracer.WithError] if 
	// [interceptor.TracerFinishSpanOptions.Error] is non-nil and not
	// [workflow.IsContinueAsNewError].
	OnFinish func(*interceptor.TracerFinishSpanOptions) []tracer.FinishOption
```

@Quinn-With-Two-Ns, preference?

File: contrib/datadog/tracing/interceptor.go
Comment (cretz) [inline contrib/datadog/tracing/interceptor.go:261]: Can just inline this, no need for a separate method (but no biggie)

**Review by cretz — COMMENTED** [review]
Looks mostly good to me, will want @Quinn-With-Two-Ns's thoughts.

File: contrib/datadog/tracing/interceptor.go
Comment (vaihtovirta) [reply contrib/datadog/tracing/interceptor.go]: Addressed

File: contrib/datadog/tracing/interceptor.go
Comment (vaihtovirta) [reply contrib/datadog/tracing/interceptor.go]: Updated the naming, will be waiting for the final decision regarding the more flexible option.

File: contrib/datadog/tracing/interceptor.go
Comment (cretz) [inline contrib/datadog/tracing/interceptor.go:254]:
```suggestion
		(t.CheckError == nil || t.CheckError(err))
```

Otherwise no error will ever be tracked unless they set the option. Alternatively you could just default check error to be a default impl that checks for continue-as-new error and tell users that so if they override, they can choose whether continue-as-new is counted.

File: contrib/datadog/tracing/interceptor.go
Comment (Quinn-With-Two-Ns) [reply contrib/datadog/tracing/interceptor.go]: +1 to using `TracerFinishSpanOptions` to allow us to add more fields in the future.

Comment (vaihtovirta) [reply contrib/datadog/tracing/interceptor.go]: Implemented 👌

File: contrib/datadog/tracing/interceptor.go
Comment (vaihtovirta) [inline contrib/datadog/tracing/interceptor.go:75]: Instead of having if-else statement in the Finish function, I've decided to set a default function in the constructor.

**Review by cretz — APPROVED** [review]
LGTM, @Quinn-With-Two-Ns?

**Review by Quinn-With-Two-Ns — APPROVED** [review]
LGTM, I think you  just need to rebase and that would fix the `feature` tests

**Pull request merged**
